gateway/api/views/files.py

```diff
@@ -67,38 +69,196 @@ def check_user_has_provider(self, user, provider_name):
-                sanitize_file_path(settings.MEDIA_ROOT),
-                sanitize_file_path(user_dir),
-            )
-            if os.path.exists(user_dir):
-                files = [
-                    os.path.basename(path)
-                    for path in glob.glob(f"{user_dir}/*.tar")
-                    + glob.glob(f"{user_dir}/*.h5")
-                ]
+                    return Response(
+                        {
+                            "message": f"Qiskit Function {provider_name}/{function_title} doesn't exist."  # pylint: disable=line-too-long
+                        },
+                        status=status.HTTP_404_NOT_FOUND,
+                    )
             else:
-                logger.warning(
-                    "Directory %s does not exist for %s.", user_dir, request.user
+                function = self.get_user_function(
+                    user=request.user, function_title=function_title
+                )
+                if not function:
+                    return Response(
+                        {"message": f"Qiskit Function {function_title} doesn't exist."},
+                        status=status.HTTP_404_NOT_FOUND,
+                    )
+
+            file_storage = FileStorage(
+                username=username,
+                working_dir=working_dir,
+                function_title=function_title,
+                provider_name=provider_name,
+            )
+            files = file_storage.get_files()
+
+        return Response({"results": files})
+
+    @action(methods=["GET"], detail=False, url_path="provider")
+    def provider_list(self, request):
+        """
+        It returns a list with the names of available files for the provider working directory:
+            provider_name/function_title
+        """
+        tracer = trace.get_tracer("gateway.tracer")
+        ctx = TraceContextTextMapPropagator().extract(carrier=request.headers)
+        with tracer.start_as_current_span("gateway.files.provider_list", context=ctx):
+            username = request.user.username
+            provider_name = sanitize_name(request.query_params.get("provider"))
+            function_title = sanitize_name(request.query_params.get("function"))
+            working_dir = PROVIDER_STORAGE
+
+            function = self.get_provider_function(
+                provider_name=provider_name, function_title=function_title
+            )
+            if not function:
+                return Response(
+                    {
+                        "message": f"Qiskit Function {provider_name}/{function_title} doesn't exist."  # pylint: disable=line-too-long
+                    },
+                    status=status.HTTP_404_NOT_FOUND,
                 )
 
+            if working_dir == PROVIDER_STORAGE:
+                if not self.user_has_provider_access(request.user, provider_name):
+                    return Response(
+                        {"message": "You don't have access to this provider."},
```

Comment:
Same here, should we hide that the provider exist? I think this is a question for product.

Reply:
Mmm... Maybe yes, as we were doing in the another comment. Let me double check it.

Reply:
Solved in [81fcfcf](https://github.com/Qiskit/qiskit-serverless/pull/1529/commits/81fcfcf87f061b76a4d4abccbe15989ddf90e897)